tasks/ast_tools/src/codegen.rs

```diff
@@ -15,16 +19,54 @@ pub struct AstCodegen {
     files: Vec<PathBuf>,
     passes: Vec<Box<dyn Runner<Output = (), Context = EarlyCtx>>>,
     generators: Vec<Box<dyn Runner<Output = GeneratorOutput, Context = LateCtx>>>,
+    derives: Vec<Box<dyn Runner<Output = DeriveOutput, Context = LateCtx>>>,
 }
 
 pub struct AstCodegenResult {
     pub schema: Schema,
-    pub outputs: Vec<(/* generator name */ &'static str, /* output */ GeneratorOutput)>,
+    pub outputs: Vec<SideEffect>,
+}
+
+pub enum SideEffect {
+    /// File write side-effect
+    Write(/* path */ PathBuf, /* output */ Vec<u8>),
+}
```

Comment:
Everything is a file write at the end of the day, I was thinking about the possibility of adding higher-level files to store data from multiple passes/generators. Something like this:

```rust
pub enum SideEffect {
    Write(/* path */ PathBuf, /* output */ Vec<u8>),
    Cache(/* key */ String, /* value */ Vec<u8>),
}
```

Having one file to store multiple outputs and separating them at runtime using their length helps us by lowering the syscalls on the reader part. But I agree that it might be unnecessary to make SideEffect an enum when we only have one variant.
I'll make it a tuple struct for now.